node/pkg/ethereum/polygonimpl.go

```diff
@@ -0,0 +1,240 @@
+	DelayInMs        int
+	logger           *zap.Logger
+	rawClient        *ethRpc.Client
+	rootClient       *ethClient.Client
+	rootCaller       *rootAbi.AbiRootChainCaller
+}
+
+var (
+	POLY_RCP_URL           int = 0
+	POLY_RCP_CONTRACT_ADDR int = 1
+)
+
+func UsePolygonRootChain(extraParams []string) bool {
+	return len(extraParams) >= 2 && extraParams[0] != "" && extraParams[1] != ""
+}
+
+func (e *PolygonImpl) SetLogger(l *zap.Logger) {
+	e.logger = l
+	e.logger.Info("RCP: using root chain proxy to detect new blocks", zap.String("eth_network", e.BaseEth.NetworkName),
+		zap.String("url", e.RootChainUrl), zap.Stringer("contract_address", e.RootChainAddress), zap.Int("delay_in_ms", e.DelayInMs))
+}
+
+func (e *PolygonImpl) DialContext(ctx context.Context, rawurl string) (err error) {
+	timeout, cancel := context.WithTimeout(ctx, 15*time.Second)
+	defer cancel()
+
+	// This is used for doing raw eth_ RPC calls.
+	e.rawClient, err = ethRpc.DialContext(timeout, rawurl)
+	if err != nil {
+		return err
+	}
+
+	// This is used for querying the Root Chain Proxy contract on Ethereum
+	e.rootClient, err = ethClient.DialContext(timeout, e.RootChainUrl)
+	if err != nil {
+		return err
+	}
+
+	// This is used for doing all other go-ethereum calls.
+	err = e.BaseEth.DialContext(ctx, rawurl)
+	return err
+}
+
+func (e *PolygonImpl) NewAbiFilterer(address ethCommon.Address) (err error) {
+	return e.BaseEth.NewAbiFilterer(address)
+}
+
+func (e *PolygonImpl) NewAbiCaller(address ethCommon.Address) (err error) {
+	e.rootCaller, err = rootAbi.NewAbiRootChainCaller(e.RootChainAddress, e.rootClient)
+	if err != nil {
+		return err
+	}
+	return e.BaseEth.NewAbiCaller(address)
+}
+
+func (e *PolygonImpl) GetCurrentGuardianSetIndex(ctx context.Context) (uint32, error) {
+	return e.BaseEth.GetCurrentGuardianSetIndex(ctx)
+}
+
+func (e *PolygonImpl) GetGuardianSet(ctx context.Context, index uint32) (ethAbi.StructsGuardianSet, error) {
+	return e.BaseEth.GetGuardianSet(ctx, index)
+}
+
+func (e *PolygonImpl) WatchLogMessagePublished(ctx, timeout context.Context, sink chan<- *ethAbi.AbiLogMessagePublished) (ethEvent.Subscription, error) {
+	return e.BaseEth.WatchLogMessagePublished(ctx, timeout, sink)
+}
+
+func (e *PolygonImpl) TransactionReceipt(ctx context.Context, txHash ethCommon.Hash) (*ethTypes.Receipt, error) {
+	return e.BaseEth.TransactionReceipt(ctx, txHash)
+}
+
+func (e *PolygonImpl) TimeOfBlockByHash(ctx context.Context, hash ethCommon.Hash) (uint64, error) {
+	return e.BaseEth.TimeOfBlockByHash(ctx, hash)
+}
+
+func (e *PolygonImpl) ParseLogMessagePublished(log ethTypes.Log) (*ethAbi.AbiLogMessagePublished, error) {
+	return e.BaseEth.ParseLogMessagePublished(log)
+}
+
+type PolygonPollSubscription struct {
+	errOnce   sync.Once
+	err       chan error
+	quit      chan error
+	unsubDone chan struct{}
+}
+
+var ErrUnsubscribedForPolygonPolling = errors.New("unsubscribed")
+
+func (sub *PolygonPollSubscription) Err() <-chan error {
+	return sub.err
+}
```

Comment:
Why not name this `Error()`? That's more in line with most Golang stuff.

Reply:
Err() is the standard name used for subscription errors in geth, so I am staying consistent.